internal/controlplane/handlers_oauth.go

```diff
@@ -115,55 +128,94 @@ func (s *Server) GetAuthorizationURL(ctx context.Context,
-// This function is used by the CLI client.
-func (s *Server) ExchangeCodeForTokenCLI(ctx context.Context,
-	in *pb.ExchangeCodeForTokenCLIRequest) (*httpbody.HttpBody, error) {
+// HandleProviderCallback handles the OAuth 2.0 authorization code callback from the enrolled
+// provider. This function gathers the state from the database and compares it to the state
+// passed in. If they match, the provider code is exchanged for a provider token.
+// note: this is an HTTP only (not RPC) handler
+func (s *Server) HandleProviderCallback() runtime.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+		ctx := r.Context()
+
+		if err := s.processCallback(ctx, w, r, pathParams); err != nil {
+			log.Printf("error handling provider callback: %s", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func (s *Server) processCallback(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	pathParams map[string]string) error {
+	state := r.URL.Query().Get("state")
+	code := r.URL.Query().Get("code")
 
 	// Configure tracing
 	span := trace.SpanFromContext(ctx)
-	span.SetName("server.ExchangeCodeForTokenCLI")
-	span.SetAttributes(attribute.Key("code").String(in.Code))
+	span.SetName("server.HandleProviderCallback")
+	span.SetAttributes(attribute.Key("code").String(code))
 	defer span.End()
 
-	// Check the nonce to make sure it's valid
-	valid, err := mcrypto.IsNonceValid(in.State, s.cfg.Auth.NoncePeriod)
+	provider := pathParams["provider"]
+	if !app.IsProviderSupported(provider) {
+		return fmt.Errorf("provider %s is not supported yet", provider)
+	}
 
+	// Check the nonce to make sure it's valid
+	valid, err := mcrypto.IsNonceValid(state, s.cfg.Auth.NoncePeriod)
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error checking nonce: %s", err)
+		return fmt.Errorf("error checking nonce: %w", err)
 	}
-
 	if !valid {
-		return nil, status.Error(codes.InvalidArgument, "invalid nonce")
+		return errors.New("invalid nonce")
 	}
 
 	// get projectID from session along with state nonce from the database
-	stateData, err := s.store.GetProjectIDPortBySessionState(ctx, in.State)
+	stateData, err := s.store.GetProjectIDPortBySessionState(ctx, state)
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error getting project ID by session state: %s", err)
+		return fmt.Errorf("error getting project ID by session state: %w", err)
 	}
 
-	// get provider
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, stateData.ProjectID)
+	err = s.generateOAuthToken(ctx, provider, code, stateData)
 	if err != nil {
-		return nil, providerError(err)
+		return err
 	}
 
-	// generate a new OAuth2 config for the given provider
-	oauthConfig, err := auth.NewOAuthConfig(provider.Name, true)
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = provider
+	logger.BusinessRecord(ctx).Project = stateData.ProjectID
```

Comment:
Can we move these before `s.generateOAuthToken, so that we include the provider and project information even on errors?